lib/src/facts/linux/filesystem_resolver.cc

```diff
@@ -41,17 +43,44 @@ namespace facter { namespace facts { namespace linux {
             LOG_ERROR("setmntent failed: %1% (%2%): mountpoints are unavailable.", strerror(errno), errno);
             return;
         }
+        string root_device;
         mntent entry;
         char buffer[4096];
         while (mntent *ptr = getmntent_r(file, &entry, buffer, sizeof(buffer))) {
+            string device = ptr->mnt_fsname;
+
             // Skip over anything that doesn't map to a device
-            if (!boost::starts_with(ptr->mnt_fsname, "/dev/")) {
+            if (!boost::starts_with(device, "/dev/")) {
                 continue;
             }
 
+            // If the "root" device, lookup the actual device from the kernel options
+            // This is done because not all systems symlink /dev/root
+            if (device == "/dev/root") {
+                if (root_device.empty()) {
+                    file::each_line("/proc/cmdline", [&](string& line) {
+                        auto begin = line.find("root=");
```

Comment:
re_search would've made this pretty succinct. I wonder what the overhead would be.